### DISCUSSION TITLE
tls: accept error: no shared cipher

### DISCUSSION THREAD

**Discussion by tdrz** [baresip/baresip#2614] [General]
I'm using baresip v3.1.0 to connect to my Twilio account. I can place calls from baresip successfully, but when I receive a call, I see the following error message:

tls: accept error: (r=-1, ssl_err=1)
tls: 139931151087424:error:1417A0C1:SSL routines:tls_post_process_client_hello:no shared cipher:../ssl/statem/statem_srvr.c:2283:

After seeing [this announcement from Twilio](https://support.twilio.com/hc/en-us/articles/360006751753-Tips-for-Upgrading-Your-Environment-to-Support-Twilio-REST-API-s-TLS-and-Strong-Cipher-Suite-Changes), I've upgraded the openssl library to the latest one (3.1.0) like this:

(this is from my Dockerfile)
```
    RUN git clone -b v3.1.0 --single-branch https://github.com/baresip/baresip.git && \
        cd baresip && \
        cmake -B build -DCMAKE_BUILD_TYPE=RelWithDebInfo -DCMAKE_C_COMPILER=clang-11 \
            -DCMAKE_CXX_COMPILER=clang++-11 -DCMAKE_INSTALL_PREFIX=/usr -DOPENSSL_ROOT_DIR=/install/openssl/openssl-3.1.0
```

But to no avail. I still get the same error, so I suspect there something else wrong.

Do you have any idea where I should be looking?

This is my baresip config file:

```
#
# baresip configuration -- example for linux
#

#------------------------------------------------------------------------------

# Core
poll_method		epoll		# poll, select, epoll ..

# SIP
#sip_listen		0.0.0.0:5060
#sip_certificate	cert.pem
#sip_cafile		/etc/ssl/certs/ca-certificates.crt
sip_transports		tls 
#udp,tcp,tls,ws,wss
#sip_trans_def		udp
sip_verify_server	no
#sip_tos			160 # See TOS fields!

## TOS fields ##
#    7     6     5     4     3     2     1     0
# +-----+-----+-----+-----+-----+-----+-----+-----+
# |   PRECEDENCE    |          TOS          | MBZ |
# +-----+-----+-----+-----+-----+-----+-----+-----+
# TOS:
# bit 4 - low delay
# bit 3 - high throughput
# bit 2 - high reliability
# bit 1 - low cost
# MBZ:
# bit 0 - zero

# Call
call_local_timeout	120
call_max_calls		4
call_hold_other_calls	yes

# Audio
#audio_path		/usr/share/baresip
audio_player		pulse
audio_source		pulse
audio_alert		pulse,speaker
#ausrc_srate		48000
#auplay_srate		48000
#ausrc_channels		0
#auplay_channels	0
#audio_txmode		poll		# poll, thread
audio_level		no
ausrc_format		s16		# s16, float, ..
auplay_format		s16		# s16, float, ..
auenc_format		s16		# s16, float, ..
audec_format		s16		# s16, float, ..
audio_buffer		20-160		# ms
audio_telev_pt		101		# payload type for telephone-event

# Video
#video_source		v4l2,/dev/video0
#video_display		x11,nil
video_size		640x480
video_bitrate		1000000
video_fps		30.00
video_fullscreen	yes
videnc_format		yuv420p

# AVT - Audio/Video Transport
rtp_tos			184
rtp_video_tos		136
#rtp_ports		10000-20000
#rtp_bandwidth		512-1024 # [kbit/s]
rtcp_mux		no
jitter_buffer_type	fixed		# off, fixed, adaptive
jitter_buffer_delay	5-10		# frames
#jitter_buffer_wish	6		# frames for start
rtp_stats		no
#rtp_timeout		60

# Network
#dns_server		1.1.1.1:53
#dns_server		1.0.0.1:53
#dns_fallback		8.8.8.8:53
#net_interface		eth0
# Play tones
#file_ausrc		aufile
#file_srate		16000
#file_channels		1

#------------------------------------------------------------------------------
# Modules

module_path		/usr/lib/x86_64-linux-gnu/baresip/modules

# UI Modules
module			stdio.so
#module			cons.so
#module			evdev.so
#module			httpd.so

# Audio codec Modules (in order)
module			opus.so
#module			amr.so
#module			g7221.so
#module			g722.so
#module			g726.so
#module			g711.so
#module			gsm.so
#module			l16.so
#module			mpa.so
#module			codec2.so
#module			ilbc.so

# Audio filter Modules (in encoding order)
module			auconv.so
module			auresamp.so
#module			vumeter.so
#module			sndfile.so
#module			plc.so
#module			webrtc_aec.so

# Audio driver Modules
#module			alsa.so
module			pulse.so
#module			jack.so
#module			portaudio.so
#module			aubridge.so
#module			aufile.so
#module			ausine.so

# Video codec Modules (in order)
#module			avcodec.so
#module			vp8.so
#module			vp9.so

# Video filter Modules (in encoding order)
#module			selfview.so
#module			snapshot.so
#module			swscale.so
#module			vidinfo.so
#module			avfilter.so

# Video source modules
#module			v4l2.so
#module			vidbridge.so

# Video display modules
#module			directfb.so
#module			x11.so
#module			sdl.so
#module			fakevideo.so

# Audio/Video source modules
#module			avformat.so
#module			gst.so
#module			gst_video.so

# Compatibility modules
#module			ebuacip.so
module			uuid.so

# Media NAT modules
module			stun.so
module			turn.so
module			ice.so
#module			natpmp.so
#module			pcp.so

# Media encryption modules
module			srtp.so
#module			dtls_srtp.so
#module			zrtp.so


#------------------------------------------------------------------------------
# Application Modules

module_app		account.so
#module_app		contact.so
#module_app		debug_cmd.so
#module_app		echo.so
#module_app		gtk.so
module_app		menu.so
#module_app		mwi.so
#module_app		presence.so
#module_app		serreg.so
#module_app		syslog.so
module_app		mqtt.so
#module_app		ctrl_tcp.so
#module_app		ctrl_dbus.so
#module_app		httpreq.so
#module_app		multicast.so
module_app		netroam.so


#------------------------------------------------------------------------------
# Module parameters

# DTLS SRTP parameters
#dtls_srtp_use_ec	prime256v1

# UI Modules parameters
cons_listen		0.0.0.0:5555 # cons - Console UI UDP/TCP sockets

http_listen		0.0.0.0:8000 # httpd - HTTP Server

ctrl_tcp_listen		0.0.0.0:4444 # ctrl_tcp - TCP interface JSON

evdev_device		/dev/input/event0

# Opus codec parameters
opus_bitrate		48000 # 6000-510000
opus_stereo		yes
#opus_sprop_stereo	yes
#opus_cbr		no
#opus_inbandfec		no
#opus_dtx		no
#opus_mirror		no
#opus_complexity	10
#opus_application	audio	# {voip,audio}
opus_samplerate	48000
#opus_packet_loss	10	# 0-100 percent (expected packet loss)

# Opus Multistream codec parameters
#opus_ms_channels	2	#total channels (2 or 4)
#opus_ms_streams	2	#number of streams
#opus_ms_c_streams	2	#number of coupled streams

vumeter_stderr		yes

#jack_connect_ports	yes

# Selfview
video_selfview		window # {window,pip}
#selfview_size		64x64

# ZRTP
#zrtp_hash		no  # Disable SDP zrtp-hash (not recommended)

# Menu
#redial_attempts	0 # Num or <inf>
#redial_delay		5 # Delay in seconds
#ringback_disabled	no
#statmode_default	off
#menu_clean_number	no
#sip_autoanswer_method	rfc5373 # {rfc5373,call-info,alert-info}
#ring_aufile		ring.wav
#callwaiting_aufile	callwaiting.wav
#ringback_aufile	ringback.wav
#notfound_aufile	notfound.wav
#busy_aufile		busy.wav
#error_aufile		error.wav
#sip_autoanswer_aufile	autoanswer.wav

# GTK
#gtk_clean_number	no

# avcodec
#avcodec_h264enc	libx264
#avcodec_h264dec	h264
#avcodec_h265enc	libx265
#avcodec_h265dec	hevc
#avcodec_hwaccel	vaapi
#avcodec_profile_level_id 42002a

# ctrl_dbus
#ctrl_dbus_use	system		# system, session

# mqtt
#mqtt_broker_host	sollentuna.example.com
#mqtt_broker_port	1883
#mqtt_broker_cafile	/path/to/broker-ca.crt	# set this to enforce TLS
#mqtt_broker_clientid	baresip01	# has to be unique
#mqtt_broker_user	user
#mqtt_broker_password	pass
#mqtt_basetopic		baresip/01

# sndfile
#snd_path		/tmp

# EBU ACIP
#ebuacip_jb_type	fixed	# auto,fixed

# HTTP request module
#httpreq_ca		trusted1.pem
#httpreq_ca		trusted2.pem
#httpreq_dns		1.1.1.1
#httpreq_dns		8.8.8.8
#httpreq_hostname	myserver
#httpreq_cert		cert.pem
#httpreq_key		key.pem

# multicast receivers (in priority order)- port number must be even
#multicast_call_prio	0
#multicast_ttl		1
#multicast_jbuf_type	fixed		# off, fixed, adaptive
#multicast_jbuf_delay	5-10		# frames
#multicast_jbuf_wish	6		# frames for start
#multicast_listener	224.0.2.21:50000
#multicast_listener	224.0.2.21:50002

# avformat
#avformat_pass_through	yes
```

**Comment by tdrz** [comment]
Weirdly enough, I sometimes do not get the error and the call can be established. But most of the time I receive the TLS error and I have no clue why.

I activated the debug_cmd module and this is the /sysinfo

--- System info: ---
 Machine:  x86_64/Linux
 Version:  3.1.0 (libre v3.1.0)
 Build:    64-bit little endian
 Kernel:   Linux baremetal01 5.10.0-22-amd64 #1 SMP Debian 5.10.178-3 (2023-04-22) x86_64
 Uptime:  18 mins 29 secs
 Started:  Sun May 28 07:16:42 2023
 Compiler: Debian Clang 11.0.1
 OpenSSL:  OpenSSL 3.1.0 14 Mar 2023

So I think I can definitely rule out the OpenSSL version.

**Comment by alfredh** [comment]
please include all relevant info, like written here:

https://github.com/baresip/baresip/wiki/Guidelines-for-reporting-Issues


Have you enabled SIP outbound, for NAT traversal?
is the SIP proxy making an inbound TCP/TLS connect to your endpoint ?


I think you have to configure your OpenSSL properly, with mentioned ciphersuites from Twilio.

**Comment by tdrz** [comment]
> I think you have to configure your OpenSSL properly, with mentioned ciphersuites from Twilio.

I did, I was able to make outbound calls successfully. Also inbond worked occasionally. 

I _think_ the problem had to do with RTP port allocation. I have multiple services that are allocated the same RTP ports pool and somehow a conflict over a port was leading to this error, imo unrelated to the actual problem.

In any case, I have isolated my baresip client in a docker container and this seems to have solved the problem.

**Comment by tdrz** [comment]
Unfortunately I am having this issue again.

> Have you enabled SIP outbound, for NAT traversal?

No, I am running baresip on a machine with a static IP. Baresip itself is running inside a kubernetes pod with host networking, so no NAT traversal is needed.

> is the SIP proxy making an inbound TCP/TLS connect to your endpoint ?

I am connecting the baresip client directly to a Twilio SIP endpoint. In the Twilio logs I can see the connection attempt (failing).

```
Unable to connect sip:+12312723789-0x558e00230de0@<my_static_ip>:33359;transport=tls
```

On the baresip client side, I can see this error **most of the time, but not always** when I make a call from the outside to baresip:

```
tls: accept error: (r=-1, ssl_err=1)
tls: C0B8C8C3307F0000:error:0A0000C1:SSL routines:tls_post_process_client_hello:no shared cipher:ssl/statem/statem_srvr.c:2220:
```
This adds to the confusion, because it sometimes works OK, I can receive a call, answer it, and the two parties can hear each other.

> I think you have to configure your OpenSSL properly, with mentioned ciphersuites from Twilio.

I did:

```
# openssl ciphers
TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:DHE-RSA-AES256-SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES256-SHA:ECDHE-RSA-AES256-SHA:DHE-RSA-AES256-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES128-SHA:DHE-RSA-AES128-SHA:RSA-PSK-AES256-GCM-SHA384:DHE-PSK-AES256-GCM-SHA384:RSA-PSK-CHACHA20-POLY1305:DHE-PSK-CHACHA20-POLY1305:ECDHE-PSK-CHACHA20-POLY1305:AES256-GCM-SHA384:PSK-AES256-GCM-SHA384:PSK-CHACHA20-POLY1305:RSA-PSK-AES128-GCM-SHA256:DHE-PSK-AES128-GCM-SHA256:AES128-GCM-SHA256:PSK-AES128-GCM-SHA256:AES256-SHA256:AES128-SHA256:ECDHE-PSK-AES256-CBC-SHA384:ECDHE-PSK-AES256-CBC-SHA:SRP-RSA-AES-256-CBC-SHA:SRP-AES-256-CBC-SHA:RSA-PSK-AES256-CBC-SHA384:DHE-PSK-AES256-CBC-SHA384:RSA-PSK-AES256-CBC-SHA:DHE-PSK-AES256-CBC-SHA:AES256-SHA:PSK-AES256-CBC-SHA384:PSK-AES256-CBC-SHA:ECDHE-PSK-AES128-CBC-SHA256:ECDHE-PSK-AES128-CBC-SHA:SRP-RSA-AES-128-CBC-SHA:SRP-AES-128-CBC-SHA:RSA-PSK-AES128-CBC-SHA256:DHE-PSK-AES128-CBC-SHA256:RSA-PSK-AES128-CBC-SHA:DHE-PSK-AES128-CBC-SHA:AES128-SHA:PSK-AES128-CBC-SHA256:PSK-AES128-CBC-SHA
```
And comparing with what [Twilio wants](https://www.twilio.com/docs/voice/api/sip-interface#tlssrtp-specifications), this seems fine.

More info as requested:

- Baresip version: I've tried multiple ones v2.6.0, v3.1.0, v3.2.0. I can see the same error for all of them
- Baresip config: see first post.
- ngrep didn't work but I ran tcpdump, don't know if that helps:
```
14:48:14.777310 IP ec2-54-172-60-3.compute-1.amazonaws.com.46179 > eb.eb.3ca9.ip4.static.sl-reverse.com.33359: Flags [S], seq 3314086951, win 29200, options [mss 1460,sackOK,TS val 3814125684 ecr 0,nop,wscale 7], length 0
14:48:14.777414 IP eb.eb.3ca9.ip4.static.sl-reverse.com.33359 > ec2-54-172-60-3.compute-1.amazonaws.com.46179: Flags [S.], seq 108568404, ack 3314086952, win 65160, options [mss 1460,sackOK,TS val 1874842000 ecr 3814125684,nop,wscale 7], length 0
14:48:14.808153 IP ec2-54-172-60-3.compute-1.amazonaws.com.46179 > eb.eb.3ca9.ip4.static.sl-reverse.com.33359: Flags [P.], seq 1:148, ack 1, win 229, options [nop,nop,TS val 3814125715 ecr 1874842000], length 147
14:48:14.808210 IP eb.eb.3ca9.ip4.static.sl-reverse.com.33359 > ec2-54-172-60-3.compute-1.amazonaws.com.46179: Flags [.], ack 148, win 508, options [nop,nop,TS val 1874842030 ecr 3814125715], length 0
14:48:14.808459 IP eb.eb.3ca9.ip4.static.sl-reverse.com.33359 > ec2-54-172-60-3.compute-1.amazonaws.com.46179: Flags [P.], seq 1:8, ack 148, win 508, options [nop,nop,TS val 1874842031 ecr 3814125715], length 7
14:48:14.808553 IP eb.eb.3ca9.ip4.static.sl-reverse.com.33359 > ec2-54-172-60-3.compute-1.amazonaws.com.46179: Flags [F.], seq 8, ack 148, win 508, options [nop,nop,TS val 1874842031 ecr 3814125715], length 0
14:48:14.839257 IP ec2-54-172-60-3.compute-1.amazonaws.com.46179 > eb.eb.3ca9.ip4.static.sl-reverse.com.33359: Flags [F.], seq 148, ack 8, win 229, options [nop,nop,TS val 3814125746 ecr 1874842031], length 0
14:48:14.839307 IP eb.eb.3ca9.ip4.static.sl-reverse.com.33359 > ec2-54-172-60-3.compute-1.amazonaws.com.46179: Flags [.], ack 149, win 508, options [nop,nop,TS val 1874842062 ecr 3814125746], length 0
14:48:14.839327 IP ec2-54-172-60-3.compute-1.amazonaws.com.46179 > eb.eb.3ca9.ip4.static.sl-reverse.com.33359: Flags [.], ack 9, win 229, options [nop,nop,TS val 3814125746 ecr 1874842031], length 0
```
The relevant traffic is the inbound one to port 33359.

Could you please tell me **how can I activate more baresip logs**, especially those relevant to SSL?

**Comment by tdrz** [comment]
I was trying to get a more detailed tcpdump but then it worked again. baresip successfully received the handshake on a different port and this time it worked. I really don't know what's going on ...

**Comment by tdrz** [comment]
OK, got a bit more data from tcpdump. There are only 2 packets (?!) in that dump, from what I understand the ciphers are not even sent by the servers and baresip already shows the error:

```
tls: accept error: (r=-1, ssl_err=1)
tls: C0B8C8C3307F0000:error:0A0000C1:SSL routines:tls_post_process_client_hello:no shared cipher:ssl/statem/statem_srvr.c:2220:
```

[trace.tar.gz](https://github.com/baresip/baresip/files/11668827/trace.tar.gz)

**Comment by cspiel1** [comment]
You could enable the key logger in `re`. Then you are able to few the tcpdump in wireshark.

```
cd re
mkdir build
cd build
cmake .. -DCMAKE_BUILD_TYPE=Debug -DTRACE_SSL=/tmp/ssl.key
```

You need to load `/tmp/ssl.key` into wireshark: `Edit - Preferences - Protocols - TLS`
In the field for "(Pre)-Master-Secret log filename" put /tmp/ssl.key

Also start baresip with option `-v`!

Even more interesting WARNINGs you get if you increase `DEBUG_LEVEL` to 6 in `re/src/tls/openssl/tls.c`. See function `tls_verify_handler()`!

**Comment by alfredh** [comment]
please enable SIP outbound to keep TCP connection alive.

If you want to run a SIP/TLS server you have to enable a certificate.


```
/sipstat
transports:
  10.0.1.137:56155 (UDP)
  10.0.1.137:56559 (TCP)
  10.0.1.137:56560 (TLS)
  10.0.1.137:9 (WS)
  10.0.1.137:9 (WSS)
  10.1.168.249:57385 (UDP)
  10.1.168.249:56573 (TCP)
  10.1.168.249:56574 (TLS)
  10.1.168.249:9 (WS)
  10.1.168.249:9 (WSS)
connections:
  [1]   TCP  10.1.168.249:56575 --> 212.79.111.155:5060  (Established)
connection configurations:
client transactions:
server transactions:
```

**Comment by tdrz** [comment]
> Also start baresip with option -v!

Yes, this I already did!

> Even more interesting WARNINGs you get if you increase DEBUG_LEVEL to 6 in re/src/tls/openssl/tls.c. See function tls_verify_handler()!

OK, will have a look!

> please enable SIP outbound to keep TCP connection alive.

I did, to no avail. I used Twilio's outbound proxy as described [here](https://www.twilio.com/blog/support-regional-sip-registration). Same tls error again after I initially was able to receive a call.

> If you want to run a SIP/TLS server you have to enable a certificate.

I had a look at running Kamailio in my k8s cluster. Unfortunately I have no experience with it and was unable to establish a connection from baresip because I need to somehow deal with Kamailio authentication.

**Comment by tdrz** [comment]
@cspiel1 

> Even more interesting WARNINGs you get if you increase DEBUG_LEVEL to 6 in re/src/tls/openssl/tls.c

I did this but I do **not** see any additional information compared to before. See my branch https://github.com/copiltembel/re/tree/debugTls

This line from my errors ```tls: accept error: (r=-1, ssl_err=1)``` seems to be output by `src/tls/openssl/tls_tcp.c` in method `tls_accept()`. 

~~I don't know where this `tls: C0B8C8C3307F0000:error:0A0000C1:SSL routines:tls_post_process_client_hello:no shared cipher:ssl/statem/statem_srvr.c:2220:` is coming from.~~

Must be from `tls_flush_error();`

My understanding is that the code fails in this method from `src/tls/openssl/tls_tcp.c`:

```
static int tls_accept(struct tls_conn *tc)
{
	int err = 0, r;

	ERR_clear_error();

	r = SSL_accept(tc->ssl); // <-- this returns an error
	if (r <= 0) {
		const int ssl_err = SSL_get_error(tc->ssl, r);

		switch (ssl_err) {

		case SSL_ERROR_WANT_READ:
			break;

		default:
			DEBUG_WARNING("accept error: (r=%d, ssl_err=%d)\n",
				      r, ssl_err);
			tls_flush_error();
			err = EPROTO;
			break;
		}

		ERR_clear_error();
	}

	return err;
}
```

**Comment by tdrz** [comment]
The way I am reproducing the issue now is by making a call from the outside to the sip number. This first call usually succeeds. Then I wait around 5 to 10 minutes and make another call. This second call fails with the tls error as described.

**Comment by tdrz** [comment]
I would really like to see more details related to the TLS error, like what ciphers were received from the servers, but I don't know how I can do that.

**Comment by alfredh** [comment]
please also include the log file:

https://github.com/baresip/baresip/wiki/Guidelines-for-reporting-Issues


Can you please try to make this experiment:

1. Setup two baresip endpoints A and B on your local network
2. Make a call from A to B
3. Observe the SIP transport `/sipstat` if there is any action TCP connection
4. Make 10-20 calls from A to B and check if works every time


This experiment excludes Twilio, btw.


The SIP stack in baresip has both a TLS client and a TLS server.
For inbound calls there are 2 options:

1. reuse existing TCP-connection (if exist)
2. make a new TCP-connection to SIP/TCP-server in baresip

when using TLS option 2.) requires a TLS certificate in baresip.

**Comment by tdrz** [comment]
> please also include the log file:

I am not sure which log **file** you are referring to, so here is the stdout/stderr :

```
$ baresip -v -T -f ~/.baresip/ -n bond1
baresip v3.2.0 Copyright (C) 2010 - 2023 Alfred E. Heggestad et al.
10:26:41.632|Local network addresses:
10:26:41.632|      bond1:  <<<MY_PUBLIC_STATIC_IP4>>>
10:26:41.632|      bond1:  <<<MY_PUBLIC_STATIC_IP6>>>
10:26:41.632|uag: add local address <<<MY_PUBLIC_STATIC_IP4>>>
10:26:41.634|uag: add local address <<<MY_PUBLIC_STATIC_IP6>>>
10:26:41.634|ui: stdio
10:26:41.636|opus: fmtp="stereo=1;sprop-stereo=1;maxaveragebitrate=28000"
10:26:41.636|aucodec: opus/48000/2
10:26:41.636|aucodec: PCMU/8000/1
10:26:41.636|aucodec: PCMA/8000/1
10:26:41.636|aufilt: auconv
10:26:41.636|aufilt: auresamp
10:26:41.636|ausrc: alsa
10:26:41.636|auplay: alsa
10:26:41.636|uuid: loaded UUID 4dec58bb-498f-24a0-b67c-3078e117ee81 from file /home/tdr/.baresip//uuid
10:26:41.636|medianat: stun
10:26:41.636|medianat: turn
10:26:41.636|medianat: ice
10:26:41.637|mediaenc: srtp
10:26:41.637|mediaenc: srtp-mand
10:26:41.637|mediaenc: srtp-mandf
10:26:41.637|module: loading app account.so
10:26:41.637|Populated 0 accounts
10:26:41.637|account: No SIP accounts found
 -- check your config or add an account using 'uanew' command
10:26:41.637|module: loading app contact.so
10:26:41.637|Populated 3 contacts
10:26:41.637|module: loading app debug_cmd.so
10:26:41.637|module: loading app menu.so
10:26:41.638|module: loading app netroam.so
10:26:41.639|Populated 3 audio codecs
10:26:41.639|Populated 2 audio filters
10:26:41.639|Populated 0 video codecs
10:26:41.639|Populated 0 video filters
10:26:41.639|baresip is ready.
/uanew <sip:<<<COMMENTED_OUT>>>@<<<SIP_SUBDOMAIN>>>.sip.twilio.com;transport=tls>;sipnat=outbound;outbound="sip:sip.ashburn.twilio.com;transport=tls";regint=600;auth_user=<<<AUTH_USER>>>;auth_pass=<<<PASSWORD>>>;mediaenc=srtp-mand;audio_codecs=opus/48000/2
Creating UA for <sip:<<<COMMENTED_OUT>>>@<<<SIP_SUBDOMAIN>>>.sip.twilio.com;transport=tls>;sipnat=outbound;outbound="sip:sip.ashburn.twilio.com;transport=tls";regint=600;auth_user=<<<AUTH_USER>>>;auth_pass=<<<PASSWORD>>>;mediaenc=srtp-mand;audio_codecs=opus/48000/2 ...
10:27:02.123|<<<COMMENTED_OUT>>>@<<<SIP_SUBDOMAIN>>>.sip.twilio.com: Using sipnat: 'outbound'
10:27:02.123|<<<COMMENTED_OUT>>>@<<<SIP_SUBDOMAIN>>>.sip.twilio.com: Using media encryption 'srtp-mand'
10:27:02.123|ua: ua_register sip:<<<COMMENTED_OUT>>>@<<<SIP_SUBDOMAIN>>>.sip.twilio.com

--- User Agents (1) ---
0 - sip:<<<COMMENTED_OUT>>>@<<<SIP_SUBDOMAIN>>>.sip.twilio.com    zzz 

10:27:03.168|<<<COMMENTED_OUT>>>@<<<SIP_SUBDOMAIN>>>.sip.twilio.com: (prio 0) {1/TLS/v4} 200 OK (Twilio) [1 binding]
All 1 useragent registered successfully! (21530 ms)
10:27:27.889|ua: sipsess connect via TLS 54.172.60.1:5061 --> <<<MY_PUBLIC_STATIC_IP4>>>:42614
10:27:27.889|<<<COMMENTED_OUT>>>@<<<SIP_SUBDOMAIN>>>.sip.twilio.com: selected for <<<COMMENTED_OUT>>>-0x557f09ba2b60
10:27:27.889|ua: using connection-address 34.203.251.120 of SDP offer
10:27:27.889|call: alloc with params laddr=<<<MY_PUBLIC_STATIC_IP4>>>, af=AF_INET, use_rtp=1
10:27:27.889|call: use_video=0
10:27:27.889|stream: audio: starting mediaenc 'srtp-mand' (wait_secure=0)
10:27:27.889|sip:<<<COMMENTED_OUT>>>@<<<SIP_SUBDOMAIN>>>.sip.twilio.com: Incoming call from: <<<REMOTE_NUMBER>>> <<<SIP_INCOMING>>><<<SIP_SUBDOMAIN>>>.sip.twilio.com - audio-video: sendrecv-inactive - (press 'a' to accept)
10:27:27.891|play: /usr/local/share/baresip/ring.wav: No such file or directory [2]
10:27:34.787|<<<SIP_INCOMING>>><<<SIP_SUBDOMAIN>>>.sip.twilio.com: session closed: Connection reset by peer [104]
10:27:34.787|audio: destroyed (started=0)
[000534870] tls: accept error: (r=-1, ssl_err=1)
[000534870] tls: C0A81B49477F0000:error:0A0000C1:SSL routines:tls_post_process_client_hello:no shared cipher:ssl/statem/statem_srvr.c:2220:
/sipstat
transports:
  <<<MY_PUBLIC_STATIC_IP4>>>:46399 (TLS)
  [<<<MY_PUBLIC_STATIC_IP6>>>]:38431 (TLS)
connections:
  [1]   TLS  <<<MY_PUBLIC_STATIC_IP4>>>:36000 --> 54.172.60.1:5061  (Established)
connection configurations:
client transactions:
server transactions:
```

I've masked the sensitive content, hope it doesn't disturb too much.

You can see the first call is successful, then the tls error occurs at the second call.

> This experiment excludes Twilio, btw.

My guess is that the problem lies in the combination of baresip and Twilio. But just a guess atm.

**Comment by tdrz** [comment]
Maybe it is too much to ask, but I could give you the credentials for one of my Twilio accounts and you could try it.

**Comment by tdrz** [comment]
>Can you please try to make this experiment:
>Setup two baresip endpoints A and B on your local network
> Make a call from A to B

Unsure how to do this. I don't need any SIP accounts? 

I could just run baresip in a docker container locally and start it twice from inside the container, but I don't know how to make that call.

**Comment by cspiel1** [comment]
> 10:27:27.889|ua: sipsess connect via TLS 54.172.60.1:5061 --> <<<MY_PUBLIC_STATIC_IP4>>>:42614

The first incoming SIP INVITE is sent to port `42614`. I guess this is the TCP src port of the SIP REGISTER. So like @alfredh wrote, the TCP connection of the REGISTER is re-used by twilio.

You wrote that there is a second incoming INVITE. And in the log we see that `tls_accept()` is called and fails. This means that baresip has now the role of an TLS server. This happens if the remote (twilio) needs to open a new TCP+TLS connection. A TLS server is normally verified. That's why baresip would need a TLS certificate.

The question is why twilio does not re-use the TCP+TLS connection of the REGISTER?

Normally I would say: Most likely the TCP connection for the REGISTER was closed already. But you configured:
- The registration interval `regint` very short, to 600 seconds.
- You configured `sipnat=outbound;outbound=...`. This enables the registration keepalive mechanism, if the SIP server supports the "outbound" extension.

So in your case maybe the keepalive mechanism is not started because the response to the REGISTER has no `Require: outbound` header and twilio has very short TCP timeouts.

**Comment by tdrz** [comment]
Thank you!

I've forwarded your answer to Twilio, let's see what comes back.

**Comment by cspiel1** [comment]
You could analyse more. Try the same with TCP transport and capture the SIP traffic, view and analyse in wireshark!

- Is there a `Require: outbound` in the 200 Ok response of the REGISTER.
- To which port the first INVITE is sent?
- To which port the second INVITE is sent?

Further: Activate the key logger: https://github.com/baresip/baresip/issues/2592#issuecomment-1580092783

Edit: Sorry, I guess it is not possible to register at twilio with SIP/TCP. Thus, I suggest you to try the key logger.

**Comment by tdrz** [comment]
> I guess it is not possible to register at twilio with SIP/TCP.

Yes, it is. I can use SIP over TCP and this seems to work fine (meaning I can receive all calls):

`/uanew <sip:<<<COMMENTED_OUT>>>@<<<SIP_SUBDOMAIN>>>.sip.twilio.com;transport=tcp>;sipnat=outbound;outbound="sip:sip.ashburn.twilio.com;transport=tcp";regint=600;auth_user=<<<AUTH_USER>>>;auth_pass=<<<PASSWORD>>>;audio_codecs=opus/48000/2`

**Comment by tdrz** [comment]
> Try the same with TCP transport and capture the SIP traffic, view and analyse in wireshark!

I need a bit more time to do this.

**Comment by tdrz** [comment]
Just as a note, for test purposes I am using a [MicroSip client](https://www.microsip.org) which is based on the PJSIP stack. I haven't had the same issue with it thus far.

MicroSip client has options for "Register Refresh" and Keep-Alive intervals (I don't know if this is PJSIP specific or MicroSip specific).

Here is a screenshot of that part of the settings:
![MicrosipSettings](https://github.com/baresip/baresip/assets/37596026/c5b25915-20fd-44f8-a69f-b7e2357b3577)

Is there a similiar "Keep-Alive" interval that I can play with in baresip?

**Comment by cspiel1** [comment]
Yes baresip/re has
- register interval, it is account parameter `;regint=`
- and such a keep alive mechanism. By default it is 120 seconds (for TCP) with a randomized part of 20%.

Thus, if you are able to register with SIP/TCP then the answer to these questions would be helpful: https://github.com/baresip/baresip/issues/2592#issuecomment-1587008212

The TLS problem most likely is a result of an unwanted TCP communication.

**Comment by tdrz** [comment]
Thank you! I'm on the road for a few days, will do it when I'm back.

**Comment by tdrz** [comment]
If I type `/timers` in the UA, I don't see the TCP/TLS keep-alive timer.

```
/timers
Timers (8):
  0x555f61d60358: th=0x7f83a76d7d44 expire=31ms file=/home/tdr/src/github/baresip/re/src/async/async.c:119
  0x555f61efa330: th=0x7f83a76ad439 expire=250ms file=/home/tdr/src/github/baresip/baresip/modules/stdio/stdio.c:89
  0x7f83a4f2e1b8: th=0x7f83a4f2b866 expire=31284ms file=/home/tdr/src/github/baresip/baresip/modules/netroam/netroam.c:206
  0x555f61f033c8: th=0x7f83a76da6e7 expire=508875ms file=/home/tdr/src/github/baresip/re/src/dns/client.c:360
  0x555f61f04738: th=0x7f83a76da6e7 expire=508875ms file=/home/tdr/src/github/baresip/re/src/dns/client.c:360
  0x555f61f04b98: th=0x7f83a76da6e7 expire=508879ms file=/home/tdr/src/github/baresip/re/src/dns/client.c:360
  0x555f61d584d0: th=0x7f83a77471a6 expire=513368ms file=/home/tdr/src/github/baresip/re/src/sipreg/reg.c:282
  0x555f61f066a0: th=0x7f83a773e950 expire=882825ms file=/home/tdr/src/github/baresip/re/src/sip/transp.c:525
```

The last one seems to be the TCP_IDLE_TIMEOUT, which for whatever reason dissapears after a while.

Note that I am using an outbound proxy with this config:

`sipnat=outbound;outbound="sip:sip.ashburn.twilio.com;transport=tls"`

Running this through the debugger, it seems that Twilio does not specify the `outbound` require header, and then baresip doesn't run a timer to keep the connection alive.

```
static void start_outbound(struct sipreg *reg, const struct sip_msg *msg)
{
	const struct sip_hdr *flowtimer;

	if (!sip_msg_hdr_has_value(msg, SIP_HDR_REQUIRE, "outbound"))
		return;

	flowtimer = sip_msg_hdr(msg, SIP_HDR_FLOW_TIMER);

	(void)sip_keepalive_start(&reg->ka, reg->sip, msg,
				  flowtimer ? pl_u32(&flowtimer->val) : 0,
				  keepalive_handler, reg);
}
```

I don't know enough about the SIP protocol yet to draw a conclusion. Any input?
